pkg/apis/networking/validation/validation_test.go

```diff
@@ -1853,3 +1853,209 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+			t.Errorf("Expected success for test '%s', got %v", k, errs)
+		}
+	}
+
+	invalidNodeSelector := makeNodeSelector("NoUppercaseOrSpecialCharsLike=Equals", api.NodeSelectorOpIn, []string{"bar"})
+
+	errorCases := map[string]*networking.ClusterCIDRConfig{
+		// Config test.
+		"nil spec.IPv4 and spec.IPv6": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(nil), setIPv6CIDR(nil)),
+		"invalid spec.NodeSelector": makeClusterCIDRConfigCustom(
+			setNodeSelector(invalidNodeSelector)),
+
+		// IPv4 tests.
+		"empty spec.IPv4.CIDR": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("", 24))),
+		"invalid spec.IPv4.CIDR": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("test", 24))),
+		"valid IPv6 CIDR in spec.IPv4.CIDR": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("fd00::/120", 24))),
+		"invalid spec.IPv4.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 100))),
+		"invalid spec.IPv4.PerNodeMaskSize < CIDR Mask": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 12))),
+
+		// IPv6 tests.
+		"empty spec.IPv6.CIDR": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("", 120))),
+		"invalid spec.IPv6.CIDR": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("testv6", 120))),
+		"valid IPv4 CIDR in spec.IPv6.CIDR": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("10.2.0.0/16", 24))),
+		"invalid spec.IPv6.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("fd00::/120", 1000))),
+		"invalid spec.IPv6.PerNodeMaskSize < CIDR Mask": makeClusterCIDRConfigCustom(
+			setIPv6CIDR(makeCIDRConfig("fd00::/120", 119))),
+
+		// DualStack test.
+		"32 - IPv4.PerNodeMaskSize != 128 - IPv6.PerNodeMaskSize": makeClusterCIDRConfigCustom(
+			setIPv4CIDR(makeCIDRConfig("10.2.0.0/16", 24)),
+			setIPv6CIDR(makeCIDRConfig("fd00::/120", 112))),
+	}
+
+	// Error cases are not expected to pass validation.
+	for testName, ccc := range errorCases {
+		if errs := ValidateClusterCIDRConfig(ccc); len(errs) == 0 {
+			t.Errorf("Expected failure for test: %s", testName)
+		}
+	}
+}
+
+func TestValidateClusterConfigUpdate(t *testing.T) {
+	oldCCC := makeValidClusterCIDRConfig()
+
+	// Tweaks used below.
+	setIPv4CIDR := func(cidrConfig *networking.CIDRConfig) cccTweak {
+		return func(ccc *networking.ClusterCIDRConfig) {
+			ccc.Spec.IPv4 = cidrConfig
+		}
+	}
+
+	setIPv6CIDR := func(cidrConfig *networking.CIDRConfig) cccTweak {
+		return func(ccc *networking.ClusterCIDRConfig) {
+			ccc.Spec.IPv6 = cidrConfig
+		}
+	}
+
+	setNodeSelector := func(nodeSelector *api.NodeSelector) cccTweak {
+		return func(ccc *networking.ClusterCIDRConfig) {
+			ccc.Spec.NodeSelector = nodeSelector
+		}
+	}
+
+	updateNodeSelector := makeNodeSelector("foo", api.NodeSelectorOpIn, []string{"bar2"})
+
+	successCases := map[string]*networking.ClusterCIDRConfig{
+		"update with no tweaks": makeClusterCIDRConfigCustom(),
```

Comment:
When this test case fails. the title won't help folks to understand what happened.

Reply:
Updated